From a23e097a7d1430688cd446297d160e5d5a9d7589 Mon Sep 17 00:00:00 2001
From: Andrew Boie <andrew.p.boie@intel.com>
Date: Thu, 26 Sep 2013 11:51:38 -0700
Subject: fastboot: bootimg: support extended command line

Fastboot has an independent implementation of mkbootimg; update it
so that the command line is properly written. We also ensure that
the array bounds are not overwritten.

In the future it will probably be better to unify these
implementations.

Issue: ADEV-148
Change-Id: I06001afd5a1e0b39bfcd05ca6a90e97a81861602
Domain: AOSP-fastboot
Category: Device Enablement
Upstream-candidate: Yes
Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
---
 fastboot/bootimg.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/fastboot/bootimg.c b/fastboot/bootimg.c
index 240784f..712996b 100644
--- a/fastboot/bootimg.c
+++ b/fastboot/bootimg.c
@@ -34,7 +34,15 @@
 
 void bootimg_set_cmdline(boot_img_hdr *h, const char *cmdline)
 {
-    strcpy((char*) h->cmdline, cmdline);
+    size_t cmdlen = strlen(cmdline);
+
+    strncpy((char *)h->cmdline, cmdline, BOOT_ARGS_SIZE - 1);
+    h->cmdline[BOOT_ARGS_SIZE - 1] = '\0';
+    if (cmdlen >= (BOOT_ARGS_SIZE - 1)) {
+        cmdline += (BOOT_ARGS_SIZE - 1);
+        strncpy((char *)h->extra_cmdline, cmdline, BOOT_EXTRA_ARGS_SIZE);
+        h->extra_cmdline[BOOT_EXTRA_ARGS_SIZE - 1] = '\0';
+    }
 }
 
 boot_img_hdr *mkbootimg(void *kernel, unsigned kernel_size, unsigned kernel_offset,
-- 
1.7.4.1

